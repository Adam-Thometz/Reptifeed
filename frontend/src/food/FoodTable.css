.FoodTable {
  table-layout: fixed;
  background-color: white;
  width: 80vw;
  margin: 0 auto;
  margin-bottom: 5%;
  border-radius: 20px;
  padding: 10px 0;
}

.FoodTable-header {
  margin: 0 1020px;
  background-color: yellow;
}

.FoodTable-header-cells {
  font-weight: 700;
}

img {
  height: 100px;
  width: 100px;
}

.FoodTable-checkbox {
  width: 20px;
  height: 20px;
}