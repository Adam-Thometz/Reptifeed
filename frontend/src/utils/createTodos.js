
/** createTodos(reptiles, pantry)
 * 
 * Takes a user's reptiles and pantry and creates a list of things they need to do to unlock the app's full features
 * 
 * Returns an two arrays of Link components, one of important things and one of nice-to-haves
 */

function createTodos(reptiles, pantry) {
  /** Essential checks:
   * - reptileCheck: Checks if a user has reptiles
   * - vegetableOftenCheck: Checks if a user has a vegetable staple in their pantry
   * - proteinOftenCheck: Checks if a user has a protein staple in their pantry
   * - treatModeratelyCheck: Checks if a user has a treat staple in their pantry
   * - supplementCheck: Checks if a user has at least two different supplements
   */
  const essentialTodos = [];

  const reptileCheck = !!reptiles.length;
  const vegetableOftenCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'often');
  const proteinOftenCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'often');
  const treatModeratelyCheck = pantry.some(food => food.isTreat && food.frequency === 'moderately');
  const supplementCheck = (pantry.filter(food => food.type === 'supplement')).length >= 2;

  if (!reptileCheck) essentialTodos.push('Add a reptile');
  if (!vegetableOftenCheck) essentialTodos.push('Add vegetables eaten often');
  if (!proteinOftenCheck) essentialTodos.push('Add protein eaten often');
  if (!treatModeratelyCheck) essentialTodos.push('Add a treat eaten moderately');
  if (!supplementCheck) essentialTodos.push('Add at least two supplements');

  /** Nice-to-have checks:
   * - vegetableModeratelyCheck: Checks if a user has a vegetable that's eaten moderately in their pantry
   * - vegetableOccasionallyCheck: Checks if a user has a vegetable that's eaten occasionally in their pantry
   * - proteinModeratelyCheck: Checks if a user has a protein that's eaten moderately in their pantry
   * - proteinOccasionallyCheck: Checks if a user has a protein that's eaten occasionally in their pantry
   * - treatOccasionallyCheck: Checks if a user has a treat that's eaten rarely in their pantry
   */
  const niceToHaveTodos = [];

  const vegetableModeratelyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'moderately');
  const vegetableOccasionallyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'occasionally' && !food.isTreat);
  const proteinModeratelyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'moderately');
  const proteinOccasionallyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'occasionally' && !food.isTreat);
  const treatOccasionallyCheck = pantry.some(food => food.isTreat && food.frequency === 'occasionally');

  if (!vegetableModeratelyCheck) niceToHaveTodos.push('Add a vegetable eaten moderately');
  if (!vegetableOccasionallyCheck) niceToHaveTodos.push('Add a vegetable eaten occasionally');
  if (!proteinModeratelyCheck) niceToHaveTodos.push('Add a protein eaten moderately');
  if (!proteinOccasionallyCheck) niceToHaveTodos.push('Add a protein eaten occasionally');
  if (!treatOccasionallyCheck) niceToHaveTodos.push('Add a treat eaten occasionally');

  return {essentialTodos, niceToHaveTodos};
};

export default createTodos;