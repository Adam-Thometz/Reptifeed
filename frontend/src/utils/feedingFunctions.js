/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
*/

function getFoodFromPantry(pantry, freq, type) {
  const availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) return `Get some more ${type}`
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  const availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) return "You don't have any treats";
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

export { getFoodFromPantry, getTreatFromPantry };