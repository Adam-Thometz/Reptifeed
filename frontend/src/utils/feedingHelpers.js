/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
 * 
 * If no foods are available for the frequency, returns foods eaten most often
*/

function getFoodFromPantry(pantry, freq, type) {
  let availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) {
    availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === 'often' && food.type === type)));
  };
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  let availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) {
    availableTreats = pantry.filter(food => food.isTreat);
  };
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

function getNextSupplement(pantry, lastSupplement) {
  const nextSupplement = lastSupplement === 'calcium' ?
    pantry.find(food => food.name === 'multivitamin') :
    pantry.find(food => food.name === 'calcium');
  
  return nextSupplement;
}

export { getFoodFromPantry, getTreatFromPantry, getNextSupplement };