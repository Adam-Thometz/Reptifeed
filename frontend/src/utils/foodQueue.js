// foodQueue: Feed reptile based on frequency
const foodQueue = ["often", "often", "moderately", "often", "often", "moderately", "often", "often", "moderately", "occasionally"];

// treatQueue: Give reptile treats based on frequency
const treatQueue = ["moderately", "moderately", "occasionally"];

export { foodQueue, treatQueue };