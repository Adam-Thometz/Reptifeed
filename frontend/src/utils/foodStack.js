// foodStack: Feed reptile based on frequency
const foodStack = ["occasionally", "moderately", "often", "often", "moderately", "often", "often", "moderately", "often", "often"];

// treatStack: Give reptile treats based on frequency
const treatStack = ["occasionally", "moderately", "moderately"];

export { foodStack, treatStack };