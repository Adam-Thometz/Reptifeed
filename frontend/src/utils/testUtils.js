import React from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: 'rickSanchez',
  password: 'ilikeMorty',
  email: 'rickman@qmail.com',
  isAdmin: false
};

const demoReptile = [
  {
    id: 1,
    name: 'two crows',
    species: 'crows',
    subspecies: 'two of them',
    birthday: '2015-10-29',
    imgUrl: '',
    ownerId: 1
  }
];

const demoPantry = [
  {
    name: 'plumbus',
    type: 'everything',
    frequency: 'always',
    image: '',
    isTreat: true,
    tips: "Everyone needs a plumbus!"
  }
];

const UserProvider = ({ children, currUser = demoUser, pantry = demoPantry, reptiles = demoReptile }) => (
  <UserContext.Provider value={{currUser, pantry, reptiles}}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };