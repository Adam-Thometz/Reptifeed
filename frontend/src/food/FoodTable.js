import React, { useContext } from "react";
import ReptifeedApi from "../api";
import UserContext from "../utils/UserContext";
import './FoodTable.css'

const FoodTable = ({ foods }) => {
  const { currUser, pantry, setPantry } = useContext(UserContext)

  const isInPantry = name => {
    return pantry.some(f => f.name === name);
  }

  const handleToggle = async food => {
    if (!isInPantry(food.name)) {
      const newFood = await ReptifeedApi.addToPantry(currUser.id, food);
      setPantry(p => [...p, newFood]);
    } else {
      await ReptifeedApi.removeFromPantry(currUser.id, food.name)
      setPantry(p => p.filter(f => f.name !== food.name));
    }
  };

  return (
    <table className="FoodTable">
      <thead className="FoodTable-header">
        <tr tabIndex={0} className="FoodTable-header-cells">
          <td>Image</td>
          <td>Name</td>
          <td>Type</td>
          <td>How often?</td>
          <td>Treat?</td>
          <td>Tips</td>
          <td>Add / Remove</td>
        </tr>
      </thead>
      <tbody className="FoodTable-body">
        {foods.map(f => (
          <tr tabIndex={0} key={f.name} className="FoodTable-body-cells">
            <td><img src={f.image} alt={f.name} /></td>
            <td>{f.name}</td>
            <td>{f.type}</td>
            <td>{f.frequency}</td>
            <td>{f.isTreat ? "Yes" : "No"}</td>
            <td>{f.tips}</td>
            <td><input className="FoodTable-checkbox" type='checkbox' checked={isInPantry(f.name)} onChange={() => handleToggle(f)} /></td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default FoodTable;

// {isInPantry(f.name) ? "Food is in pantry" : <button onClick={() => handleUpdate(f)}>Add to Pantry</button>}