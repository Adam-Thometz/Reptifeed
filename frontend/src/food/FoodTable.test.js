import React from "react";
import FoodTable from "./FoodTable";
import { render, UserProvider } from '../utils/testUtils'

const food = [
  {
    name: 'plumbus',
    type: 'everything',
    frequency: 'always',
    image: '',
    isTreat: true,
    tips: "Everyone needs a plumbus!"
  }
];

test('it renders without crashing', () => {
  render(<UserProvider>
    <FoodTable foods={food} />
  </UserProvider>);
});

test('it matches snapshot', () => {
  const { asFragment } = render(
    <UserProvider>
      <FoodTable foods={food} />
    </UserProvider>);
  expect(asFragment()).toMatchSnapshot();
});