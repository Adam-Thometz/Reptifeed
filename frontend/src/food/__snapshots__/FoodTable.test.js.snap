// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="FoodTable"
  >
    <thead
      class="FoodTable-header"
    >
      <tr
        class="FoodTable-header-cells"
      >
        <td>
          Image
        </td>
        <td>
          Name
        </td>
        <td>
          Type
        </td>
        <td>
          How often?
        </td>
        <td>
          Treat?
        </td>
        <td>
          Tips
        </td>
        <td>
          Add / Remove
        </td>
      </tr>
    </thead>
    <tbody
      class="FoodTable-body"
    >
      <tr
        class="FoodTable-body-cells"
      >
        <td>
          <img
            alt="plumbus"
            src=""
          />
        </td>
        <td>
          plumbus
        </td>
        <td>
          everything
        </td>
        <td>
          always
        </td>
        <td>
          Yes
        </td>
        <td>
          Everyone needs a plumbus!
        </td>
        <td>
          <input
            checked=""
            class="FoodTable-checkbox"
            type="checkbox"
          />
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
